Exchanging JSON web token with an access token failed: Input required and not supplied: oidc-provider-name

### Describe the bug

env variables are not recognized - regression in v3.5.0

githubenv variables are not recognized, causing fallback to OICD which fails with this error

### Current behavior

2024-01-17T17:02:34.2039427Z ::group::Setup JFrog CLI
2024-01-17T17:02:34.2039979Z ##[group]Setup JFrog CLI
2024-01-17T17:02:34.2056995Z ##[debug]Searching for JF_URL
2024-01-17T17:02:34.2057524Z ##[debug]JF_URL found
2024-01-17T17:02:34.2058097Z ##[debug]Searching for JF_ACCESS_TOKEN, JF_USER and JF_PASSWORD
2024-01-17T17:02:34.2059057Z JF_ACCESS_TOKEN and JF_USER + JF_PASSWORD weren't found. Getting access token using OpenID Connect
2024-01-17T17:02:34.2059835Z ##[debug]Fetching JSON web token
2024-01-17T17:02:34.2061755Z ##[debug]ID token url is https://pipelinesghubeus26.actions.githubusercontent.com/k75H1DBAfn7O6Mo0TRQmb014cDWVwCbkrZ6RY4JoockeD1mW9W/00000000-0000-0000-0000-000000000000/_apis/distributedtask/hubs/Actions/plans/a5b3151a-44ff-4c0a-bb81-7544446f9994/jobs/6604c111-019b-5ed5-ae03-3fa9f8925512/idtoken?api-version=2.0
2024-01-17T17:02:34.3042451Z ::add-mask::***
2024-01-17T17:02:34.3045055Z ##[debug]Exchanging JSON web token with an access token
2024-01-17T17:02:34.3070577Z ##[error]Exchanging JSON web token with an access token failed: Input required and not supplied: oidc-provider-name
2024-01-17T17:02:34.3071739Z ::endgroup::
2024-01-17T17:02:34.3072010Z ##[endgroup]
2024-01-17T17:02:34.3123024Z ##[debug]Node Action run completed with exit code 1
2024-01-17T17:02:34.3125937Z ##[debug]JFROG_CLI_ENV_EXCLUDE='password;secret;key;token;auth;JF_ARTIFACTORY_;JF_ENV_;JF_URL;JF_USER;JF_PASSWORD;JF_ACCESS_TOKEN'
2024-01-17T17:02:34.3126916Z ##[debug]JFROG_CLI_OFFER_CONFIG='false'
2024-01-17T17:02:34.3127418Z ##[debug]JFROG_CLI_BUILD_NAME='goldenfrog-demo'
2024-01-17T17:02:34.3127921Z ##[debug]JFROG_CLI_BUILD_NUMBER='50'
2024-01-17T17:02:34.3128783Z ##[debug]JFROG_CLI_BUILD_URL='***********'
2024-01-17T17:02:34.3130110Z ##[debug]JFROG_CLI_USER_AGENT='setup-jfrog-cli-github-action/3.5.0'

### Reproduction steps

setup jfrog cli using setup action v3.5.0

### Expected behavior

no error

### Setup JFrog CLI version

3.5.0

### JFrog CLI version

2.52.9

### Workflow operating system type and version

linux, ubuntu 

### JFrog Artifactory version (if relevant)

_No response_

### JFrog Xray version (if relevant)

_No response_

--- a/.github/workflows/oidc-test.yml
+++ b/.github/workflows/oidc-test.yml
@@ -54,7 +54,7 @@ jobs:
                 "name": "oidc-test-identity-mapping",
                 "priority": "1",
                 "claims": {
-                  "sub": "repo:jfrog/setup-jfrog-cli:ref:refs/heads/main",
+                  "sub": "repo:${{ github.repository_owner }}/setup-jfrog-cli:ref:${{ github.ref }}",
                   "iss": "https://token.actions.githubusercontent.com"
                 },
                 "token_spec": {
@@ -72,7 +72,7 @@ jobs:
 
       - name: Test JFrog CLI
         run: |
-          jf rt ping
+          jf rt s "some-repo/"
 
       # Removing the OIDC integration will remove the Identity Mapping as well
       - name: Delete OIDC integration

--- a/lib/utils.js
+++ b/lib/utils.js
@@ -79,16 +79,17 @@ class Utils {
      * @returns true if OpenID Connect authentication should be used
      */
     static shouldUseOpenIDConnect(jfrogCredentials) {
+        if (!process.env.ACTIONS_ID_TOKEN_REQUEST_URL) {
+            // To enable OpenIDConnect authentication, users must configure the 'id-token: write' permission, which sets the ACTIONS_ID_TOKEN_REQUEST_URL environment variable.
+            // If this variable is empty, it indicates that OIDC should not be utilized.
+            return false;
+        }
         if (!jfrogCredentials.jfrogUrl) {
             // If no JFrog URL is specified, we can't use OpenID Connect
             return false;
         }
         if (jfrogCredentials.password || jfrogCredentials.accessToken) {
             // If credentials are specified - use them instead
-            return false;
-        }
-        if (!process.env.ACTIONS_ID_TOKEN_REQUEST_URL) {
-            // Only if the user added the 'id-token: write' permission, this environment variable was set
             return false;
         }
         return true;

--- a/lib/utils.js
+++ b/lib/utils.js
@@ -51,7 +51,7 @@ class Utils {
     static getJfrogCredentials() {
         return __awaiter(this, void 0, void 0, function* () {
             let jfrogCredentials = this.collectJfrogCredentialsFromEnvVars();
-            if (!jfrogCredentials.jfrogUrl || jfrogCredentials.password || jfrogCredentials.accessToken || !process.env.ACTIONS_ID_TOKEN_REQUEST_URL) {
+            if (!this.shouldUseOpenIDConnect(jfrogCredentials)) {
                 // Use JF_ENV or the credentials found in the environment variables
                 return jfrogCredentials;
             }
@@ -72,6 +72,26 @@ class Utils {
                 throw new Error(`Exchanging JSON web token with an access token failed: ${error.message}`);
             }
         });
+    }
+    /**
+     * Returns true if OpenID Connect authentication should be used.
+     * @param jfrogCredentials - Credentials retrieved from the environment variables
+     * @returns true if OpenID Connect authentication should be used
+     */
+    static shouldUseOpenIDConnect(jfrogCredentials) {
+        if (!jfrogCredentials.jfrogUrl) {
+            // If no JFrog URL is specified, we can't use OpenID Connect
+            return false;
+        }
+        if (jfrogCredentials.password || jfrogCredentials.accessToken) {
+            // If credentials are specified - use them instead
+            return false;
+        }
+        if (!process.env.ACTIONS_ID_TOKEN_REQUEST_URL) {
+            // Only if the user added the 'id-token: write' permission, this environment variable was set
+            return false;
+        }
+        return true;
     }
     /**
      * Gathers JFrog's credentials from environment variables and delivers them in a JfrogCredentials structure

--- a/lib/utils.js
+++ b/lib/utils.js
@@ -51,14 +51,11 @@ class Utils {
     static getJfrogCredentials() {
         return __awaiter(this, void 0, void 0, function* () {
             let jfrogCredentials = this.collectJfrogCredentialsFromEnvVars();
-            if (!jfrogCredentials.jfrogUrl) {
+            if (!jfrogCredentials.jfrogUrl || jfrogCredentials.password || jfrogCredentials.accessToken || !process.env.ACTIONS_ID_TOKEN_REQUEST_URL) {
+                // Use JF_ENV or the credentials found in the environment variables
                 return jfrogCredentials;
             }
-            // If the required credentials, such as the access token or a combination of username and password, are available, the process terminates without triggering the OIDC flow
-            if (jfrogCredentials.accessToken || (jfrogCredentials.username && jfrogCredentials.password)) {
-                return jfrogCredentials;
-            }
-            core.info("JF_ACCESS_TOKEN and JF_USER + JF_PASSWORD weren't found. Getting access token using OpenID Connect");
+            core.info('The JFrog platform credentials were not configured. Obtaining an access token through OpenID Connect.');
             const audience = core.getInput(Utils.OIDC_AUDIENCE_ARG);
             let jsonWebToken;
             try {
@@ -79,23 +76,20 @@ class Utils {
     /**
      * Gathers JFrog's credentials from environment variables and delivers them in a JfrogCredentials structure
      * @returns JfrogCredentials struct with all credentials found in environment variables
+     * @throws Error if a password provided without a username
      */
     static collectJfrogCredentialsFromEnvVars() {
-        let jfrogCredentials = {};
-        core.debug('Searching for JF_URL');
-        if (process.env.JF_URL) {
-            core.debug('JF_URL found');
-            jfrogCredentials.jfrogUrl = process.env.JF_URL;
+        let jfrogCredentials = {
+            jfrogUrl: process.env.JF_URL,
+            accessToken: process.env.JF_ACCESS_TOKEN,
+            username: process.env.JF_USER,
+            password: process.env.JF_PASSWORD,
+        };
+        if (jfrogCredentials.password && !jfrogCredentials.username) {
+            throw new Error('JF_PASSWORD is configured, but the JF_USER environment variable was not set.');
         }
-        core.debug('Searching for JF_ACCESS_TOKEN, JF_USER and JF_PASSWORD');
-        if (process.env.JF_ACCESS_TOKEN) {
-            core.debug('JF_ACCESS_TOKEN found');
-            jfrogCredentials.accessToken = process.env.JF_ACCESS_TOKEN;
-        }
-        if (process.env.JF_USER && process.env.JF_PASSWORD) {
-            core.debug('JF_USER and JF_PASSWORD found');
-            jfrogCredentials.username = process.env.JF_USER;
-            jfrogCredentials.password = process.env.JF_PASSWORD;
+        if (jfrogCredentials.username && !jfrogCredentials.accessToken && !jfrogCredentials.password) {
+            throw new Error('JF_USER is configured, but the JF_PASSWORD or JF_ACCESS_TOKEN environment variables were not set.');
         }
         return jfrogCredentials;
     }
@@ -125,6 +119,9 @@ class Utils {
             const responseString = yield response.readBody();
             const responseJson = JSON.parse(responseString);
             jfrogCredentials.accessToken = responseJson.access_token;
+            if (jfrogCredentials.accessToken) {
+                core.setSecret(jfrogCredentials.accessToken);
+            }
             return jfrogCredentials;
         });
     }

--- a/src/utils.ts
+++ b/src/utils.ts
@@ -50,16 +50,12 @@ export class Utils {
      */
     public static async getJfrogCredentials(): Promise<JfrogCredentials> {
         let jfrogCredentials: JfrogCredentials = this.collectJfrogCredentialsFromEnvVars();
-        if (!jfrogCredentials.jfrogUrl) {
+        if (!jfrogCredentials.jfrogUrl || jfrogCredentials.password || jfrogCredentials.accessToken || !process.env.ACTIONS_ID_TOKEN_REQUEST_URL) {
+            // Use JF_ENV or the credentials found in the environment variables
             return jfrogCredentials;
         }
 
-        // If the required credentials, such as the access token or a combination of username and password, are available, the process terminates without triggering the OIDC flow
-        if (jfrogCredentials.accessToken || (jfrogCredentials.username && jfrogCredentials.password)) {
-            return jfrogCredentials;
-        }
-
-        core.info("JF_ACCESS_TOKEN and JF_USER + JF_PASSWORD weren't found. Getting access token using OpenID Connect");
+        core.info('The JFrog platform credentials were not configured. Obtaining an access token through OpenID Connect.');
         const audience: string = core.getInput(Utils.OIDC_AUDIENCE_ARG);
         let jsonWebToken: string | undefined;
         try {
@@ -79,25 +75,21 @@ export class Utils {
     /**
      * Gathers JFrog's credentials from environment variables and delivers them in a JfrogCredentials structure
      * @returns JfrogCredentials struct with all credentials found in environment variables
+     * @throws Error if a password provided without a username
      */
     public static collectJfrogCredentialsFromEnvVars(): JfrogCredentials {
-        let jfrogCredentials: JfrogCredentials = {} as JfrogCredentials;
-        core.debug('Searching for JF_URL');
-        if (process.env.JF_URL) {
-            core.debug('JF_URL found');
-            jfrogCredentials.jfrogUrl = process.env.JF_URL;
-        }
+        let jfrogCredentials: JfrogCredentials = {
+            jfrogUrl: process.env.JF_URL,
+            accessToken: process.env.JF_ACCESS_TOKEN,
+            username: process.env.JF_USER,
+            password: process.env.JF_PASSWORD,
+        } as JfrogCredentials;
 
-        core.debug('Searching for JF_ACCESS_TOKEN, JF_USER and JF_PASSWORD');
-        if (process.env.JF_ACCESS_TOKEN) {
-            core.debug('JF_ACCESS_TOKEN found');
-            jfrogCredentials.accessToken = process.env.JF_ACCESS_TOKEN;
+        if (jfrogCredentials.password && !jfrogCredentials.username) {
+            throw new Error('JF_PASSWORD is configured, but the JF_USER environment variable was not set.');
         }
-
-        if (process.env.JF_USER && process.env.JF_PASSWORD) {
-            core.debug('JF_USER and JF_PASSWORD found');
-            jfrogCredentials.username = process.env.JF_USER;
-            jfrogCredentials.password = process.env.JF_PASSWORD;
+        if (jfrogCredentials.username && !jfrogCredentials.accessToken && !jfrogCredentials.password) {
+            throw new Error('JF_USER is configured, but the JF_PASSWORD or JF_ACCESS_TOKEN environment variables were not set.');
         }
         return jfrogCredentials;
     }
@@ -130,6 +122,9 @@ export class Utils {
         const responseString: string = await response.readBody();
         const responseJson: TokenExchangeResponseData = JSON.parse(responseString);
         jfrogCredentials.accessToken = responseJson.access_token;
+        if (jfrogCredentials.accessToken) {
+            core.setSecret(jfrogCredentials.accessToken);
+        }
         return jfrogCredentials;
     }
 

--- a/src/utils.ts
+++ b/src/utils.ts
@@ -78,16 +78,17 @@ export class Utils {
      * @returns true if OpenID Connect authentication should be used
      */
     private static shouldUseOpenIDConnect(jfrogCredentials: JfrogCredentials): boolean {
+        if (!process.env.ACTIONS_ID_TOKEN_REQUEST_URL) {
+            // To enable OpenIDConnect authentication, users must configure the 'id-token: write' permission, which sets the ACTIONS_ID_TOKEN_REQUEST_URL environment variable.
+            // If this variable is empty, it indicates that OIDC should not be utilized.
+            return false;
+        }
         if (!jfrogCredentials.jfrogUrl) {
             // If no JFrog URL is specified, we can't use OpenID Connect
             return false;
         }
         if (jfrogCredentials.password || jfrogCredentials.accessToken) {
             // If credentials are specified - use them instead
-            return false;
-        }
-        if (!process.env.ACTIONS_ID_TOKEN_REQUEST_URL) {
-            // Only if the user added the 'id-token: write' permission, this environment variable was set
             return false;
         }
         return true;

--- a/src/utils.ts
+++ b/src/utils.ts
@@ -50,7 +50,7 @@ export class Utils {
      */
     public static async getJfrogCredentials(): Promise<JfrogCredentials> {
         let jfrogCredentials: JfrogCredentials = this.collectJfrogCredentialsFromEnvVars();
-        if (!jfrogCredentials.jfrogUrl || jfrogCredentials.password || jfrogCredentials.accessToken || !process.env.ACTIONS_ID_TOKEN_REQUEST_URL) {
+        if (!this.shouldUseOpenIDConnect(jfrogCredentials)) {
             // Use JF_ENV or the credentials found in the environment variables
             return jfrogCredentials;
         }
@@ -70,6 +70,27 @@ export class Utils {
         } catch (error: any) {
             throw new Error(`Exchanging JSON web token with an access token failed: ${error.message}`);
         }
+    }
+
+    /**
+     * Returns true if OpenID Connect authentication should be used.
+     * @param jfrogCredentials - Credentials retrieved from the environment variables
+     * @returns true if OpenID Connect authentication should be used
+     */
+    private static shouldUseOpenIDConnect(jfrogCredentials: JfrogCredentials): boolean {
+        if (!jfrogCredentials.jfrogUrl) {
+            // If no JFrog URL is specified, we can't use OpenID Connect
+            return false;
+        }
+        if (jfrogCredentials.password || jfrogCredentials.accessToken) {
+            // If credentials are specified - use them instead
+            return false;
+        }
+        if (!process.env.ACTIONS_ID_TOKEN_REQUEST_URL) {
+            // Only if the user added the 'id-token: write' permission, this environment variable was set
+            return false;
+        }
+        return true;
     }
 
     /**

--- a/test/main.spec.ts
+++ b/test/main.spec.ts
@@ -82,7 +82,7 @@ describe('Collect credentials from environment variables test', () => {
     ];
 
     test.each(cases)(
-        'Checking Jfrog credentials struct for url: %s, access token %s, username: %s, password: %s',
+        'Checking JFrog credentials struct for url: %s, access token %s, username: %s, password: %s',
         (jfrogUrl, accessToken, username, password) => {
             process.env['JF_URL'] = jfrogUrl;
             process.env['JF_ACCESS_TOKEN'] = accessToken;
@@ -93,53 +93,43 @@ describe('Collect credentials from environment variables test', () => {
             if (jfrogUrl) {
                 expect(jfrogCredentials.jfrogUrl).toEqual(jfrogUrl);
             } else {
-                expect(jfrogCredentials.jfrogUrl).toBeUndefined();
+                expect(jfrogCredentials.jfrogUrl).toBeFalsy();
             }
 
             if (accessToken) {
                 expect(jfrogCredentials.accessToken).toEqual(accessToken);
             } else {
-                expect(jfrogCredentials.accessToken).toBeUndefined();
+                expect(jfrogCredentials.accessToken).toBeFalsy();
             }
 
             if (username) {
                 expect(jfrogCredentials.username).toEqual(username);
             } else {
-                expect(jfrogCredentials.username).toBeUndefined();
+                expect(jfrogCredentials.username).toBeFalsy();
             }
 
             if (password) {
                 expect(jfrogCredentials.password).toEqual(password);
             } else {
-                expect(jfrogCredentials.password).toBeUndefined();
+                expect(jfrogCredentials.password).toBeFalsy();
             }
         },
     );
 });
 
-test('Collect JFrog Credentials from env vars', async () => {
-    process.env['JF_URL'] = '';
-    let jfrogCredentials: JfrogCredentials = Utils.collectJfrogCredentialsFromEnvVars();
-    expect(jfrogCredentials.jfrogUrl).toBeUndefined();
-    expect(jfrogCredentials.username).toBeUndefined();
-    expect(jfrogCredentials.password).toBeUndefined();
-    expect(jfrogCredentials.accessToken).toBeUndefined();
+describe('Collect JFrog Credentials from env vars exceptions', () => {
+    let cases: string[][] = [
+        // [JF_USER, JF_PASSWORD, EXCEPTION]
+        ['', 'password', 'JF_PASSWORD is configured, but the JF_USER environment variable was not set.'],
+        ['user', '', 'JF_USER is configured, but the JF_PASSWORD or JF_ACCESS_TOKEN environment variables were not set.'],
+    ];
 
-    process.env['JF_URL'] = 'https://my-server.io';
-    process.env['JF_ACCESS_TOKEN'] = 'my-access-token';
-    jfrogCredentials = Utils.collectJfrogCredentialsFromEnvVars();
-    expect(jfrogCredentials.jfrogUrl).toEqual('https://my-server.io');
-    expect(jfrogCredentials.username).toBeUndefined();
-    expect(jfrogCredentials.password).toBeUndefined();
-    expect(jfrogCredentials.accessToken).toEqual('my-access-token');
-
-    process.env['JF_USER'] = 'user';
-    process.env['JF_PASSWORD'] = 'password';
-    jfrogCredentials = Utils.collectJfrogCredentialsFromEnvVars();
-    expect(jfrogCredentials.jfrogUrl).toEqual('https://my-server.io');
-    expect(jfrogCredentials.username).toEqual('user');
-    expect(jfrogCredentials.password).toEqual('password');
-    expect(jfrogCredentials.accessToken).toEqual('my-access-token');
+    test.each(cases)('Checking JFrog credentials struct for username: %s, password: %s', (username, password, exception) => {
+        process.env['JF_ACCESS_TOKEN'] = '';
+        process.env['JF_USER'] = username;
+        process.env['JF_PASSWORD'] = password;
+        expect(() => Utils.collectJfrogCredentialsFromEnvVars()).toThrow(new Error(exception));
+    });
 });
 
 test('Get separate env config', async () => {

